### ISSUE TITLE
`graphene-sgx helloworld` failed within docker container

### ISSUE THREAD

**Issue by sammyne** [oscarlab/graphene#2408]
## Description of the problem

Following the [building instructions](https://graphene.readthedocs.io/en/latest/building.html#id1) within docker container failed with following logs, 

```bash
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
error: Unexpected value in host CPUID. Exiting...
```

The docker image is built using this [Dockerfile](https://github.com/sammyne/graphene-testbot/blob/main/Dockerfile).

## Steps to reproduce

```bash
docker build -t graphene-testbot:alpha .

docker run --rm \
        --device /dev/kmsg:/dev/kmsg    \
        --device /dev/gsgx:/dev/gsgx    \
        --device /dev/sgx:/dev/sgx      \
        graphene-testbot:alpha
```

## Expected results

Okay with no error~

## Actual results

```bash
error: Using insecure argv source. Graphene will continue application execution, but this configuration must not be used in production!
error: Unexpected value in host CPUID. Exiting...
```

## Additional information
- os: Ubuntu 18.04.5 LTS
- docker: 18.09.9, build 039a7df9ba
- sgx driver: 1.36

Really appreciate if someone could help me out~

**Comment by dimakuv** [comment]
Graphene thinks that CPUID leaf is "bad" and loudly fails. The check happens here: https://github.com/oscarlab/graphene/blob/86ba92e962dcee023ff55fa04b90930c37e0102d/Pal/src/host/Linux-SGX/db_misc.c#L274.

Is it *only* inside the Docker container? What happens if you run Graphene on your host?

In general, a surprising error. What kind of CPU do you have?

**Comment by sammyne** [comment]
> Graphene thinks that CPUID leaf is "bad" and loudly fails. The check happens here:
> 
> https://github.com/oscarlab/graphene/blob/86ba92e962dcee023ff55fa04b90930c37e0102d/Pal/src/host/Linux-SGX/db_misc.c#L274
> 
> .
> Is it _only_ inside the Docker container? What happens if you run Graphene on your host?
> 
> In general, a surprising error. What kind of CPU do you have?

Copying files out of the built docker containers and running it on host doesn't help, either~

`uname -a` produces 

```bash
Linux ubuntu 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

CPU model name is Intel(R) Xeon(R) Platinum 8374C.

**Comment by dimakuv** [comment]
Hm, looks like you have a normal Intel Cooper Lake CPU with nothing fancy...

Unfortunately, you'll have to debug this Graphene failure. I recommend you to build and install Graphene in debug mode and debug it with GDB. Please see instructions here: https://graphene.readthedocs.io/en/latest/building.html and https://graphene.readthedocs.io/en/latest/building.html#additional-build-options. For debugging with GDB, please see instructions here: https://graphene.readthedocs.io/en/latest/devel/debugging.html#debugging-with-sgx-support.

You can set a breakpoint on `sanity_check_cpuid` function and step through it to understand what condition exactly fails.

**Comment by sammyne** [comment]
Great thanks for the tip~

I found the error occurred in case of `MPX_1` with the false `if` at line 278.

https://github.com/oscarlab/graphene/blob/86ba92e962dcee023ff55fa04b90930c37e0102d/Pal/src/host/Linux-SGX/db_misc.c#L265-L286

Specifically, the `values` passed in goes as 

```bash
eax=0x00000040 ebx=0x00000440 ecx=0x00000000 edx=0x00000000
```

hitting the 13th leaf as follows

```bash
root@ubuntu:~# cpuid -r -1 | grep -n3 'eax=0x00000040 ebx=0x00000440 ecx=0x00000000 edx=0x00000000' 
19-   0x0000000d 0x00: eax=0x000002e7 ebx=0x00000a88 ecx=0x00000a88 edx=0x00000000
20-   0x0000000d 0x01: eax=0x0000000f ebx=0x00000988 ecx=0x00000100 edx=0x00000000
21-   0x0000000d 0x02: eax=0x00000100 ebx=0x00000240 ecx=0x00000000 edx=0x00000000
22:   0x0000000d 0x05: eax=0x00000040 ebx=0x00000440 ecx=0x00000000 edx=0x00000000
23-   0x0000000d 0x06: eax=0x00000200 ebx=0x00000480 ecx=0x00000000 edx=0x00000000
24-   0x0000000d 0x07: eax=0x00000400 ebx=0x00000680 ecx=0x00000000 edx=0x00000000
25-   0x0000000d 0x08: eax=0x00000080 ebx=0x00000000 ecx=0x00000001 edx=0x00000000
```

However, I think the SGX related `values` should be read from the 7th leaf, which has `eax` set as 0 correctly.

Any idea of helping me out of this?
@dimakuv

**Comment by dimakuv** [comment]
@sammyne Awesome, thanks for this detailed analysis! From what I see, this is a bug in Graphene: we put this redundant check on line 278. This check should be removed from Graphene. Let me prepare a simple PR to fix this.

**Comment by dimakuv** [comment]
@thomasknauth Do you maybe remember why you added this check? See https://github.com/oscarlab/graphene/blob/86ba92e962dcee023ff55fa04b90930c37e0102d/Pal/src/host/Linux-SGX/db_misc.c#L278-L281.

It looks unneeded. It means: "If the SGX enclave did not request to use a particular CPU XSAVE feature, verify that this feature is not available on this CPU". But this doesn't make sense: the CPU can have more XSAVE features than the SGX enclave wants to use (and thus specifies in its XFRM).

I think the correct fix is to *not verify* but instead *force to zero* the EAX value. I'll prepare a PR now.

**Comment by dimakuv** [comment]
@sammyne Could you try https://github.com/oscarlab/graphene/pull/2409 and report if this fixes your issue?

**Comment by sammyne** [comment]
> @sammyne Could you try #2409 and report if this fixes your issue?

It works! Thank you so much~😄 @dimakuv

**Comment by sammyne** [comment]
This issue is closed thanks to the fix #2409

**Comment by boryspoplawski** [comment]
@sammyne Please don't close issues manually, it will be closed once we merge #2409 (which did not happen yet).
